### PR TITLE
doc: update json2

### PR DESCRIPTION
**Pull request by enghitalo**
> The name `json2` was chosen to avoid any unwanted potential conflicts with the
> existing codegen tailored for the main `json` module which is powered by CJSON.

`x.json2` is an experimental JSON parser written from scratch on V.

#### encode[T]

```v
import x.json2
import time

struct Person {
mut:
	name     string
	age      ?int = 20
	birthday time.Time
	deathday ?time.Time = none
}

fn main() {
	mut person := Person{
		name: 'Bob'
		birthday: time.now()
	}
	person_json := json2.encode[Person](person)
	// person_json == {"name": "Bob", "age": 20, "birthday": "2022-03-11T13:54:25.000Z"}
}
```

#### decode[T]

```v
import x.json2
import time

struct Person {
mut:
	name     string
	age      ?int = 20
	birthday time.Time
	deathday ?time.Time = none
}

fn main() {
	resp := '{"name": "Bob", "age": 20, "birthday": ${time.now()}}'
	person := json2.decode[Person](resp)!
	/*
	struct Person {
      mut:
          name "Bob"
          age  20
          birthday "2022-03-11 13:54:25"
      }
	*/
}
```
decode[T] is smart and can auto convert the type of struct fields this means that all
bellow will have the same result

```v ignore
json2.decode[Person]('{"name": "Bob", "age": 20, "birthday": "2022-03-11T13:54:25.000Z"}')!
json2.decode[Person]('{"name": "Bob", "age": 20, "birthday": "2022-03-11 13:54:25.000"}')!
json2.decode[Person]('{"name": "Bob", "age": "20", "birthday": 1647006865}')!
json2.decode[Person]('{"name": "Bob", "age": "20", "birthday": "1647006865"}}')!
```

#### raw decode

```v
import x.json2
import net.http

fn main() {
	resp := http.get('https://reqres.in/api/products/1')!

	// This return a Any type
	raw_product := json2.raw_decode(resp.body)!
}
```
#### Casting `Any` type / Navigating
```v
import x.json2
import net.http

fn main() {
	resp := http.get('https://reqres.in/api/products/1')!

	raw_product := json2.raw_decode(resp.body)!

	product := raw_product.as_map()
	data := product['data'] as map[string]json2.Any

	id := data['id'].int() // 1
	name := data['name'].str() // cerulean
	year := data['year'].int() // 2000
}
```
#### Constructing an `Any` type
```v
import x.json2

fn main() {
	mut me := map[string]json2.Any{}
	me['name'] = 'Bob'
	me['age'] = 18

	mut arr := []json2.Any{}
	arr << 'rock'
	arr << 'papers'
	arr << json2.null
	arr << 12

	me['interests'] = arr

	mut pets := map[string]json2.Any{}
	pets['Sam'] = 'Maltese Shitzu'
	me['pets'] = pets

	// Stringify to JSON
	println(me.str())
	//{
	//   "name":"Bob",
	//   "age":18,
	//   "interests":["rock","papers","scissors",null,12],
	//   "pets":{"Sam":"Maltese"}
	//}
}
```
### Null Values
`x.json2` has a separate `null` type for differentiating an undefined value and a null value.
To verify that the field you're accessing is a `null`, use `[typ] is json2.Null`.

```v ignore
fn (mut p Person) from_json(f json2.Any) {
    obj := f.as_map()
    if obj['age'] is json2.Null {
        // use a default value
        p.age = 10
    }
}
```

## Casting a value to an incompatible type
`x.json2` provides methods for turning `Any` types into usable types.
The following list shows the possible outputs when casting a value to an incompatible type.

1. Casting non-array values as array (`arr()`) will return an array with the value as the content.
2. Casting non-map values as map (`as_map()`) will return a map with the value as the content.
3. Casting non-string values to string (`str()`) will return the
JSON string representation of the value.
4. Casting non-numeric values to int/float (`int()`/`i64()`/`f32()`/`f64()`) will return zero.

### CODE REVIEW

**Comment by Delta456** [comment]
What else is needed to be done in `x.json2`? If all work is done then we can remove `cjson` and make `x.json2` the default json encoder/decoder.

**Comment by enghitalo** [comment]
> What else is needed to be done in `x.json2`? If all work is done then we can remove `cjson` and make `x.json2` the default json encoder/decoder.
@Delta456 
json2 is so close to be ready.
Basically, What is miss is whats is in `*_todo_test.vv` files

In resume: 
    for encode: Struct alias, sumtypes and Struct references
    for decode:optionals, Struct alias, sumtypes and Struct references

some problems with arrays, too.

**Comment by enghitalo** [comment]
@Delta456 Today, json2 use `field.unaliased_typ` to comparation. I will need study a bit more about `final_sym` to see a best solution

File: vlib/x/json2/README.md [hunk 1]
```diff
@@ -46,86 +120,8 @@ fn main() {
...
-```v ignore
-fn (mut p Person) from_json(f json2.Any) {
-    mp := an.as_map()
-	mut js_field_name := ''
-    $for field in Person.fields {
-        js_field_name = field.name
-
-        for attr in field.attrs {
-			if attr.starts_with('json:') {
-				js_field_name = attr.all_after('json:').trim_left(' ')
```
Comment (spytheman) [inline vlib/x/json2/README.md:114]: do we have now support for `json:` attributes built in `x.json2` ?

File: vlib/x/json2/README.md [hunk 2]
```diff
@@ -4,24 +4,98 @@
...
 fn main() {
-	// Decoding
-	resp := http.get('https://example.com')!
+	resp := '{"name": "Bob", "age": 20, "birthday": "2022-03-11T13:54:25.000Z"}'
 
-	// raw decode
-	raw_person := json2.raw_decode(resp.body)!
+	// This return a Any type
+	raw_person := json2.raw_decode(resp)!
+}
```
[the remark this answers is not in the document]
Comment (enghitalo) [reply vlib/x/json2/README.md:78]: What about `https://reqres.in/api/products`?

**Comment by ylluminate** [comment]
@Delta456 there will need to still be a significant amount of performance tuning before the `x.json2` can replace `cjson` and/or `cjson` be removed altogether. If you review @spytheman's benchmark addition in context of current comparative speeds he has observed that it's currently 10x slower without `-prod` and still 5x slower with `-prod`, etc.

File: vlib/x/json2/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (spytheman) [reply vlib/x/json2/README.md:78]: > https://reqres.in/api/products

The URL can be different, yes - as long as the overall code structure is preserved, that shows how to get the json text, then decode it using `json2.raw_decode`, then say print the result.

**Comment by enghitalo** [comment]
> @Delta456 there will need to still be a significant amount of performance tuning before the `x.json2` can replace `cjson` and/or `cjson` be removed altogether. If you review @spytheman's benchmark addition in context of current comparative speeds he has observed that it's currently 10x slower without `-prod` and still 5x slower with `-prod`, etc.

The best performative result for now, is using `-no-bounds-checking` to compile.

👇From spytheman macbook


This project will need be refactored soon.

In fact, the project needs to be refactored right now. but I'm practically not having time to contribute to it. At these end of year parties, I will have even less time. The project is very easy to finish. Very easy to refactor also. An experienced v developer with time would probably finish all the missing features in a maximum of two or three days.

File: vlib/x/json2/README.md [hunk 3]
```diff
@@ -4,24 +4,103 @@
...
 import net.http
 
 fn main() {
-	// Decoding
-	resp := http.get('https://example.com')!
+	resp := http.get('https://reqres.in/api/products/1')!
 
-	// raw decode
-	raw_person := json2.raw_decode(resp.body)!
+	// This return a Any type
```
Comment (JalonSolov) [inline vlib/x/json2/README.md:77]:
```suggestion
	// This returns an Any type
```

File: vlib/x/json2/README.md [hunk 4]
```diff
@@ -46,86 +125,8 @@ fn main() {
...
-			'age' { u.age = mp[js_field_name].int() }
-			'pets' { u.pets = mp[js_field_name].arr().map(it.str()) }
-			else {}
-		}
-    }
-}
-```
-
 ### Null Values
 `x.json2` has a separate `null` type for differentiating an undefined value and a null value.
```
Comment (JalonSolov) [inline vlib/x/json2/README.md:131]:
```suggestion
`x.json2` has a separate `Null` type for differentiating an undefined value and a null value.
```

File: vlib/x/json2/README.md [hunk 5]
```diff
@@ -46,86 +125,8 @@ fn main() {
...
 To verify that the field you're accessing is a `null`, use `[typ] is json2.Null`.
```
Comment (JalonSolov) [inline vlib/x/json2/README.md:132]:
```suggestion
To verify that the field you're accessing is a `Null`, use `[typ] is json2.Null`.
```

**Comment by spytheman** [comment]
The freebsd build failure is unrelated.

**Pull request merged**